pyiron_base/generic/flattenedstorage.py

```diff
@@ -364,6 +364,102 @@ def has_array(self, name):
+
+        >>> store.apply("double", np.sum, inplace=False, per="chunk")
+        array([ 0,  2,  6, 12])
+        >>> store.has_array("sum(double)") # returns None
+
+        In any case the return value will be the flattened array in storage so it is mostly useful for applying
+        functions per chunk, since applying per element you won't know the chunk boundaries.
+
+        >>> store.apply("integers", lambda n: 2*n, inplace=False, per="element")
+        array([0, 0, 2, 0, 2, 4, 0, 2, 4, 6])
+
+        Note that it makes no sense to apply a function per element to a per chunk array.
+
+        Args:
+            name (str): name of the array to apply the function to
+            function (Function): any callable, return value must have the same length along the first axis as the array
+                                 specified by name
+            inplace (bool/str, optional): if str, save the result in an array in this storage under the given name; if
+                                          True generate a name automatically from name and function; if False just
+                                          return the result
+            per (str, optional): apply the given function either per element or per chunk
+
+        Returns:
+            :class:`numpy.ndarray`: if per element; first axis cooresponds to the number of elements stored; if per
+                                    chunk to the number of chunks; in any case the remaining axes are determined by the
+                                    result of the function
+
+        Raises:
+            :class:`ValueError`: per is not "element" or "chunk"
+            :class:`ValueError`: if per element and the return value of function does not match the number of elements
+                                 in the first axis
+        """
+        if per == "element":
+            source = self._per_element_arrays[name][:self.num_elements]
+            result = function(source)
+            if result.shape[0] != source.shape[0]:
+                raise ValueError(f"Shape of result ({result.shape[0]}) doesn't match shape of source ({source.shape[0]})!")
+            result.resize((self.num_elements_alloc,) + result.shape[1:])
+            store = self._per_element_arrays
+        elif per == "chunk":
+            if name in self._per_element_arrays:
+                source = self._per_element_arrays[name][:self.num_elements]
+                result = np.array([ function(source[self._get_per_element_slice(i)]) for i in range(len(self)) ])
+            else:
+                source = self._per_chunk_arrays[name][:self.num_chunks]
+                result = np.array([ function(source[i]) for i in range(len(self)) ])
+            result.resize((self.num_chunks_alloc,) + result.shape[1:])
+            store = self._per_chunk_arrays
+        else:
+            raise ValueError(f"per must \"element\" or \"chunk\", not {per}")
+
+        if inplace:
```

Comment:
If this save in this case? Ok, we probably do not want to support `0` or `''` as name. However, wouldn't a `1` cause problems?
```suggestion
        if inplace is True or isinstance(inplace, str) and len(inplace)>0:
```
I just realize that it is indeed modified inplace if `name` and `inplace` match. Do we want this?
This would also have very nasty side effects for `'length'` e.g. as result for a norm function!

Reply:
Agree, this should check if the target name already exists in the storage.

Reply:
`0` or `1` would make problems indeed, but the array names are explicitly documented as strings.  I will add code to enforce this.